ebpf/src/processeventssubscriber.cpp

```diff
@@ -0,0 +1,182 @@
+
+osquery::Status ProcessEventsSubscriber::create(
+    IEventSubscriberRef& subscriber) {
+  try {
+    auto ptr = new ProcessEventsSubscriber();
+    subscriber.reset(ptr);
+
+    return osquery::Status(0);
+
+  } catch (const std::bad_alloc&) {
+    return osquery::Status(1, "Memory allocation failure");
+
+  } catch (const osquery::Status& status) {
+    return status;
+  }
+}
+
+ProcessEventsSubscriber::~ProcessEventsSubscriber() {}
+
+osquery::Status ProcessEventsSubscriber::initialize() noexcept {
+  return osquery::Status(0);
+}
+
+void ProcessEventsSubscriber::release() noexcept {}
+
+osquery::Status ProcessEventsSubscriber::configure(
+    ProcessEventsPublisher::SubscriptionContextRef subscription_context,
+    const json11::Json&) noexcept {
+  subscription_context->system_call_filter.insert(__NR_execve);
+  subscription_context->system_call_filter.insert(__NR_execveat);
+  subscription_context->system_call_filter.insert(KPROBE_FORK_CALL);
+  subscription_context->system_call_filter.insert(KPROBE_VFORK_CALL);
+  subscription_context->system_call_filter.insert(KPROBE_CLONE_CALL);
+
+  return osquery::Status(0);
+}
+
+osquery::Status ProcessEventsSubscriber::callback(
+    osquery::QueryData& new_data,
+    ProcessEventsPublisher::SubscriptionContextRef subscription_context,
+    ProcessEventsPublisher::EventContextRef event_context) {
+  new_data = {};
+
+  for (const auto& event : event_context->probe_event_list) {
+    osquery::Row row = {};
+    row["timestamp"] = std::to_string(event.timestamp / 1000U);
+    row["ppid"] = std::to_string(event.parent_tgid);
+    row["pid"] = std::to_string(event.tgid);
+    row["tid"] = std::to_string(event.pid);
+    row["uid"] = std::to_string(event.uid);
+    row["gid"] = std::to_string(event.gid);
+    row["event"] = getSystemCallName(event.function_identifier);
+
+    std::string docker_container_id = {};
+    auto status =
+        getProbeEventField(docker_container_id, event, "docker_container_id");
+
+    if (!status.ok()) {
+      row["docker_container_id"] = "";
+    } else {
+      row["docker_container_id"] = docker_container_id;
+    }
+
+    std::string exit_code = {};
+    if (event.exit_code) {
+      exit_code = std::to_string(event.exit_code.get());
+    }
+
+    row["exit_code"] = exit_code;
+
+    if (event.function_identifier == __NR_execve ||
+        event.function_identifier == __NR_execveat) {
+      std::string filename;
+      status = getProbeEventField(filename, event, "filename");
+      if (!status.ok()) {
+        row["filename"] = "";
+      } else {
+        row["filename"] = filename;
+      }
+
+      ProbeEvent::StringList argv;
+      status = getProbeEventField(argv, event, "argv");
+      if (!status.ok()) {
+        row["argv"] = "";
+
+      } else {
+        const auto& argument_list = argv.data;
+
+        std::stringstream buffer;
+
+        for (const auto& argument : argument_list) {
+          if (!buffer.str().empty()) {
+            buffer << ", ";
```

Comment:
Maybe construct strict CSV here instead? One of its redeeming qualities is that you can quote all fields to avoid escaping `,`, and `"` literals just become `""` 😄